File: python/timestep/utils/hashing.py
"""Hashing utilities."""

import hashlib
import json
from typing import Any


def stable_hash(obj: Any) -> str:
    """Generate stable hash of JSON-serializable object."""
    s = json.dumps(obj, sort_keys=True, ensure_ascii=False)
    return hashlib.sha256(s.encode("utf-8")).hexdigest()[:12]

File: typescript/timestep/utils/hashing.ts
/** Hashing utilities. */

import { createHash } from 'crypto';

export function stableHash(obj: any): string {
  const s = JSON.stringify(obj, Object.keys(obj).sort());
  return createHash('sha256').update(s, 'utf-8').digest('hex').slice(0, 12);
}
